derp/derp_server.go

```diff
@@ -944,6 +1028,48 @@ func (s *Server) notePeerSendLocked(src key.NodePublic, dst *sclient) {
 	m[dst.key] = dst.connNum
 }
 
+// getFlowLocked allocates a flowStats to track bytes/packets sent in
+// the flow identified by this dstKey and srcOrFwd. It should only be called when
+// the flowStat for this connection does not yet exist.
+func (c *sclient) getFlowStatsLocked(dstKey, srcOrFwd key.NodePublic) *flowStats {
+	s := c.s
+	// Allocate the map for all flows associated with this sclient
+	// if necessary
+	if _, ok := s.flows[c]; !ok {
+		s.flows[c] = make(map[key.NodePublic]map[key.NodePublic]*flowStats)
+	}
+	// Allocate the map for all flows to this destination if necessary
+	if _, ok := s.flows[c][dstKey]; !ok {
+		s.flows[c][dstKey] = make(map[key.NodePublic]*flowStats)
+	}
+
+	if _, ok := s.flows[c][dstKey][srcOrFwd]; ok {
+		s.logf("[unexpected] flowStat already allocated for s.flows[%v(%v)][%v][%v]",
+			c.key.ShortString(), c.connNum, dstKey.ShortString(), srcOrFwd.ShortString())
+	}
+	f := flowStats{}
+	s.flows[c][dstKey][srcOrFwd] = &f
+	// Populate the sclient's flow cache outside the lock
+	return &f
+}
+
+// updateFlowStats tracks every byte that enters the derp for a
+// src,dst,connNum tuple. f must be the valid flow, either looked up from the src
+// client's flows or recently allocated and stored only in the server
+// flow map.
+func (c *sclient) updateFlowStats(f *flowStats, dstKey, srcOrFwd key.NodePublic, contents []byte) {
+	if _, ok := c.flows[dstKey]; !ok {
+		c.flows[dstKey] = make(map[key.NodePublic]*flowStats)
```

Comment:
https://pkg.go.dev/tailscale.com/util/mak#Set may be helpful here and in a few other places

Reply:
Thanks! I was able to use mak.Set() in the initial allocation but it didn't make sense here where the common case is lookup.